nextcloud/spreed
# Fix media automatically enabled after selecting a device during a call

Call flags may not be updated to the right values without #7013

Since defe61742 when a media device (microphone or camera) is selected during a call in which the user joined without media of that kind the media is automatically enabled. This is an unexpected behaviour for the user, so even if a media device is selected it should be kept disabled until explicitly enabled by the user.

However, just reverting the commit to explicitly disable again the media is not a valid approach, as in some scenarios the event handler could disable again the media that was being enabled even before the other event handlers were executed (thus causing them to handle first the disabled event and then the enabled event), which ends messing up the state of the objects.

For example, when joining a call with video (without having joined any other call first in that session) the sent video would be a black stream, and as soon as another participant joins the video would be automatically disabled.

Therefore, rather than reverting the commit, this pull request changes the LocalMediaModel to base its enabled state on the state of the tracks (so the UI always reflects the actual track state rather than the expected track state) and removes the restriction of enabling and disabling the media only when available (which ensures that if media is not available it will be initially disabled once a device is selected).

Besides that it also ensures that media will be disabled when starting a call without microphone nor camera, and fixes some other minor things to make the code more robust.

Note that this pull request does not address the problem of media getting disabled when switching to a different device. That behaviour is a side effect of disabling the media when not available, as during the device change the media becomes temporary unavailable. It would make sense to keep the media enabled when switching to a different device and only disabling it if it is actually unavailable (either because no device was selected or because there was an error when getting the media), but this is something for another pull request.

The scenarios below can be performed with or without HPB. Nevertheless, they have been tested ad nauseam, so trying any other creative scenario would be better to find wrong behaviours that may have slipped :-)

## How to test (scenario 1)

- Start a call with audio and video
- Enable both audio and video (if they were already enabled, disable and enable them again)
- Leave the call
- Reload the page
- Start the call again, this time without audio nor video
- Open the device settings and select a microphone or a camera

### Result with this pull request

The UI shows the media as disabled, and `OCA.Talk.SimpleWebRTC.webrtc.localStreams[0].getTracks()[0].enabled` is false

### Result without this pull request

The UI shows the media as disabled, but `OCA.Talk.SimpleWebRTC.webrtc.localStreams[0].getTracks()[0].enabled` is true

## How to test (scenario 1b)

- Start a call with audio and video
- Enable both audio and video (if they were already enabled, disable and enable them again)
- Leave the call
- Start the call again, this time without audio nor video
- Open the device settings and select a microphone or a camera

### Result with this pull request

The UI shows the media as disabled, and `OCA.Talk.SimpleWebRTC.webrtc.localStreams[0].getTracks()[0].enabled` is false

### Result without this pull request

The UI shows the media as disabled, but `OCA.Talk.SimpleWebRTC.webrtc.localStreams[0].getTracks()[0].enabled` is true



## How to test (scenario 2)

- Start a call with audio and video
- Enable both audio and video (if they were already enabled, disable and enable them again)
- Leave the call
- Start the call again, this time without audio nor video
- In a private window, join the call also without audio nor video
- In the original window, open the device settings and select a microphone or a camera

### Result with this pull request

The UI shows the media as disabled, and `OCA.Talk.SimpleWebRTC.webrtc.peers[0].pc.getSenders()[0].track` is not set

### Result without this pull request

The UI shows the media as enabled, and `OCA.Talk.SimpleWebRTC.webrtc.peers[0].pc.getSenders()[0].track` is set


## How to test (scenario 2b)

- Start a call with audio and video
- Enable both audio and video (if they were already enabled, disable and enable them again)
- Leave the call
- Start the call again, this time without audio nor video; before unselecting the devices, disable audio and video
- In a private window, join the call also without audio nor video
- Open the device settings and select a microphone or a camera

### Result with this pull request

The UI shows the media as disabled, and `OCA.Talk.SimpleWebRTC.webrtc.peers[0].pc.getSenders()[0].track` is not set

### Result without this pull request

The UI shows the media as enabled, and `OCA.Talk.SimpleWebRTC.webrtc.peers[0].pc.getSenders()[0].track` is set



## How to test (scenario 3)

- Start a call with audio and video
- Disable audio, video and background blur (if they were already disabled, enable and disable them again)
- Leave the call
- Reload the page
- Start the call again without audio nor video
- In a private window, join the call also without audio nor video
- In the original window, open the device settings and select a microphone or (preferably, to check the background blur) a camera

### Result with this pull request

The UI shows the media as disabled, and `OCA.Talk.SimpleWebRTC.webrtc.peers[0].pc.getSenders()[0].track` is not set; background blur is disabled and it will not be visible if video is enabled

### Result without this pull request

The UI shows the media as enabled, and `OCA.Talk.SimpleWebRTC.webrtc.peers[0].pc.getSenders()[0].track` is set; background blur is also enabled and it is visible



## How to test (scenario 4 - Mimics what would happen if a [renegotiation](https://github.com/nextcloud/spreed/pull/6896) happened 30 seconds after the last participant joined and the messages to send the current media state to other participants are no longer being sent)

- Comment the code that [sends the current media state to other participants](https://github.com/nextcloud/spreed/blob/43ebbd09a2b9139601b945a1ab77fc95a30a5290/src/utils/webrtc/webrtc.js#L207-L216)
- Start a call with audio and video
- Enable both audio and video (if they were already enabled, disable and enable them again)
- Leave the call
- Start the call again, this time without audio nor video
- In a private window, join the call also without audio nor video
- In the original window, open the device settings and select a microphone or a camera

### Result with this pull request

The UI shows the media as disabled, and `OCA.Talk.SimpleWebRTC.webrtc.peers[0].pc.getSenders()[0].track` is not set

### Result without this pull request

The UI shows the media as disabled, but `OCA.Talk.SimpleWebRTC.webrtc.peers[0].pc.getSenders()[0].track` is set. Moreover, the participant in the private window is able to hear or see the participant in the original window!



## How to test (scenario 5)

- Start a call with audio and video
- Enable both audio and video (if they were already enabled, disable and enable them again)
- Leave the call
- Reload the page
- Start the call again without audio nor video
- Open the device settings and select a microphone or a camera
- Leave the call
- Start a call again

### Result with this pull request

In the device checker the audio or video is disabled (as audio and video were disabled in the previous call)

### Result without this pull request

In the device checker the audio or video is enabled
